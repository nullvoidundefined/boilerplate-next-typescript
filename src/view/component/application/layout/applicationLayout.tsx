import { ReactNode, useState } from "react";
import { Container } from "react-bootstrap";
import styled from "styled-components";

import { ApplicationFooter } from "../footer/applicationFooter";
import { ApplicationHeader } from "../header/applicationHeader";
import { ApplicationNavigationDropdown } from "../navigation/dropdown/applicationNavigationDropdown";
import { ApplicationNavigationSidebar } from "../navigation/sidebar/applicationNavigationSidebar";

import { COLOR, DIMENSION } from "../../../../constant";
import { User } from "../../../../type";

interface ContentProps {
    heightOffset: number;
}

interface ApplicationLayoutProps {
    children: ReactNode;
    contentHeightOffset: number;
    isMobile: boolean;
    navigationDropdownHeightOffset: number;
    onLogInButtonClick: () => void;
    onLogOutButtonClick: () => void;
    onLogoClick: () => void;
    user: User | null;
}

const StyledApplicationLayout = styled.div`
    min-width: ${DIMENSION.MINIMUM_WIDTH}px;
`;

const ApplicationContent = styled.main<ContentProps>`
    background-color: ${COLOR.CONTENT_BACKGROUND};
    min-height: ${({ heightOffset }) => {
        return `calc(100vh - ${heightOffset}px)`;
    }};
`;

const ApplicationLayout = ({
    children,
    contentHeightOffset,
    isMobile,
    navigationDropdownHeightOffset,
    onLogInButtonClick,
    onLogOutButtonClick,
    onLogoClick,
    user,
}: ApplicationLayoutProps) => {
    const [isNavigationDropdownOpen, setIsNavigationDropdownOpen] =
        useState(false);

    const toggleDropdown = () => {
        setIsNavigationDropdownOpen(!isNavigationDropdownOpen);
    };

    return (
        <StyledApplicationLayout>
            <ApplicationHeader
                isAuth={Boolean(user)}
                isMobile={isMobile}
                isNavigationDropdownOpen={isNavigationDropdownOpen}
                onDropDownToggleButtonClick={toggleDropdown}
                onLogInButtonClick={onLogInButtonClick}
                onLogOutButtonClick={onLogOutButtonClick}
                onLogoClick={onLogoClick}
            />
            {isMobile ? (
                <div className="position-relative">
                    <ApplicationNavigationDropdown
                        heightOffset={navigationDropdownHeightOffset}
                        isOpen={isNavigationDropdownOpen}
                        onLinkClick={toggleDropdown}
                    />
                    <ApplicationContent heightOffset={contentHeightOffset}>
                        {children}
                    </ApplicationContent>
                </div>
            ) : (
                <div className="d-flex">
                    <ApplicationNavigationSidebar
                        heightOffset={contentHeightOffset}
                    />
                    <div className="w-100">
                        <ApplicationContent heightOffset={contentHeightOffset}>
                            <Container>{children}</Container>
                        </ApplicationContent>
                    </div>
                </div>
            )}
            <ApplicationFooter />
        </StyledApplicationLayout>
    );
};

export { ApplicationLayout };
