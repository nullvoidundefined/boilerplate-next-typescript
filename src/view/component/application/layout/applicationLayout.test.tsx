import "jest-styled-components";

import { ApplicationLayout } from "./applicationLayout";
import renderer from "react-test-renderer";
import { noop } from "lodash";

it("renders correctly", () => {
    const props = {
        children: "Hello Layout!",
        contentHeightOffset: 0,
        isMobile: false,
        navigationDropdownHeightOffset: 0,
        onLogInButtonClick: () => noop(),
        onLogOutButtonClick: () => noop(),
        onLogoClick: () => noop(),
        user: null,
    };
    const component = renderer
        .create(<ApplicationLayout {...props} />)
        .toJSON();
    expect(component).toMatchSnapshot();
});
