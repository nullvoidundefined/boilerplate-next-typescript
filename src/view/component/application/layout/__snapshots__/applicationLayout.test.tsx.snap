// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c6 {
  background-color: blue;
  height: 120px;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 40px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: red;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 80px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 10px;
}

.c4 {
  background-color: green;
  max-width: 160px;
  min-height: calc(100vh - 0px);
  width: 160px;
}

.c0 {
  min-width: 390px;
}

.c5 {
  min-height: calc(100vh - 0px);
}

<div
  className="c0"
>
  <header
    className="c1"
  >
    <h1>
      LOGO
    </h1>
    <div
      className="c2"
    >
      <button
        className="c3 btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span>
          Log In
        </span>
      </button>
    </div>
  </header>
  <div
    className="d-flex"
  >
    <div
      className="c4"
    >
      Sidebar
    </div>
    <div
      className="w-100"
    >
      <main
        className="c5"
      >
        <div
          className="container"
        >
          Hello Layout!
        </div>
      </main>
    </div>
  </div>
  <footer
    className="c6"
  >
    Footer
  </footer>
</div>
`;
